src/runtime/refiner.ts

```diff
@@ -956,3 +956,138 @@ export class SQLExtracter {
     return `SELECT * FROM ${table}` + (filterTerms.length ? ` WHERE ${filterTerms.join(' AND ')}` : '') + ';';
   }
 }
+
+export class Fraction {
+  num: Polynomial;
+  den: Polynomial;
+
+  constructor(n?: Polynomial, d?: Polynomial) {
+    this.num = n ? Polynomial.copyOf(n) : new Polynomial();
+    this.den = d ? Polynomial.copyOf(d) : new Polynomial([1]);
+    if (this.den.isZero()) {
+      throw new Error('Division by zero.');
+    }
+    this.reduce();
+  }
+
+  static add(a: Fraction, b: Fraction): Fraction {
+    const den = Polynomial.multiply(a.den, b.den);
+    const num = Polynomial.add(Polynomial.multiply(a.num, b.den), Polynomial.multiply(b.num, a.den));
+    return new Fraction(num, den);
+  }
+
+  static negate(a: Fraction): Fraction {
+    return new Fraction(Polynomial.negate(a.num), a.den);
+  }
+
+  static subtract(a: Fraction, b: Fraction): Fraction {
+    const negB = Fraction.negate(b);
+    return Fraction.add(a, negB);
+  }
+
+  static multiply(a: Fraction, b: Fraction): Fraction {
+    return new Fraction(Polynomial.multiply(a.num, b.num), Polynomial.multiply(a.den, b.den));
+  }
+
+  static divide(a: Fraction, b: Fraction): Fraction {
+    const invB = new Fraction(b.den, b.num);
+    return Fraction.multiply(a, invB);
+  }
+
+  reduce() {
+    if (this.num.isZero()) {
+      this.den = new Polynomial([1]);
```

Comment:
Should we be concerned about the `.den` being zero here?

Reply:
The constructor already checks for ```.den``` being zero before calling ```reduce()```. It should not be possible to construct a fraction with zero ```.den``` in the first place (unless someone explicitly maliciously sets it to zero).

Reply:
Might be worth doing, ideally we'd put the requirement that it can't be 0 in the type system and not have to worry about it, but we'll just have to use runtime checking of our invariants.